### DISCUSSION TITLE
Packaging with MOOSE python modules as dependencies

### DISCUSSION THREAD

**Discussion by jvwilliams23** [idaholab/moose#29085] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi,

I have made a small python package for doing UQ on workflows including MOOSE simulations. To make my work usable, I want to package it up with `pip` and incorporate some unit testing etc. Some parts of the code depend on MOOSE's python modules `pyhit` and `moosetree` for reading, modifying and writing input files for simulations. As far as I can tell, the only way to use `pyhit` and `moosetree` is to add them to my `PYTHONPATH` manually.

I was wondering, is there any way around this? For running CI unit tests, I do not really want to add MOOSE as a submodule and clone it for all the tests, since it is a pretty large repo and I only need a tiny part of it. As a worst-case scenario, I could just avoid writing tests for that part of the code.

**Comment by GiudGiud** [comment]
Hello

There used to be a moose-tools package you could get from conda. 
@milljm will know more where we stand on this

**Reply by milljm** [reply]
Are you looking for the `moose-pyhit` package?

```bash
$> conda create -n testing moose-pyhit=2022
$> conda activate testing
$> python
```

```python
>>> from pyhit import Node
>>>
```

_Note: I am specifically informing you to install an older version (2022). The latest is not working, which I am attempting to fix: #29090._

**Reply by jvwilliams23** [reply]
@milljm Yeah I think so. Also `moosetree`. But I cannot seem to install it on my machine (mac). Get various errors. Also on our linux hpc, I get an error that glibc is too old (even though I .

```
conda create -v -n testing moose-pyhit=2022

Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working...
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed
Traceback (most recent call last):
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/exceptions.py", line 1118, in __call__
    return func(*args, **kwargs)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/cli/main.py", line 69, in main_subshell
    exit_code = do_call(args, p)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/cli/conda_argparse.py", line 91, in do_call
    return getattr(module, func_name)(args, parser)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/notices/core.py", line 109, in wrapper
    return func(*args, **kwargs)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/cli/main_create.py", line 41, in execute
    install(args, parser, 'create')
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/cli/install.py", line 307, in install
    raise e
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/cli/install.py", line 264, in install
    unlink_link_transaction = solver.solve_for_transaction(
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/core/solve.py", line 136, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/core/solve.py", line 179, in solve_for_diff
    final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/core/solve.py", line 303, in solve_final_state
    ssc = self._run_sat(ssc)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/common/io.py", line 84, in decorated
    return f(*args, **kwds)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/core/solve.py", line 848, in _run_sat
    ssc.solution_precs = ssc.r.solve(tuple(final_environment_specs),
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/common/io.py", line 84, in decorated
    return f(*args, **kwds)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/resolve.py", line 1326, in solve
    self.find_conflicts(specs, specs_to_add, history_specs)
  File "/Users/josh.williams/miniconda3/lib/python3.10/site-packages/conda/resolve.py", line 354, in find_conflicts
    raise UnsatisfiableError(bad_deps, strict=strict_channel_priority)
conda.exceptions.UnsatisfiableError

```
$ conda create -n testing python=3.10 moose-pyhit=2022

Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working...
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

Output in format: Requested package -> Available versions

Package python conflicts for:
moose-pyhit=2022 -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.10
moose-pyhit=2022 -> setuptools[version='>=65.5,<65.6.0a0'] -> python[version='>=3.11,<3.12.0a0']
```

on linux machine (even though I am able to run pyhit on this machine when on my PYTHONPATH).

```conda create -n testing moose-pyhit=2022
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: |
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.17=0
  - moose-pyhit=2022 -> libgcc-ng[version='>=11.2.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.17
```

**Reply by milljm** [reply]
What channels are you operating with? Its possibly your version of Conda may be out of date as well, as I see its _not_ using the libmamba solver (or maybe it is, kinda hard to tell in the above output). Can you run the following and report back?:

```bash
conda info
```

**Reply by jvwilliams23** [reply]
Output below from the HPC. Note I use MOOSE built from source, not with mamba.

```
          conda version : 23.3.1
    conda-build version : 3.24.0
         python version : 3.10.9.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.17=0
                          __linux=3.10.0=0
                          __unix=0=0
       base environment : /lustre/scafellpike/local/HT04544/sht09/jxw92-sht09/anaconda3  (writable)
      conda av data dir : /lustre/scafellpike/local/HT04544/sht09/jxw92-sht09/anaconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.software.inl.gov/public/linux-64
                          https://conda.software.inl.gov/public/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
```

mac channels output
```
           channel URLs : https://conda.software.inl.gov/public/osx-arm64
                          https://conda.software.inl.gov/public/noarch
                          https://repo.anaconda.com/pkgs/main/osx-arm64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-arm64
                          https://repo.anaconda.com/pkgs/r/noarch
```

**Reply by milljm** [reply]
Yeah, it looks like you're using an older instance of Conda. Not the issue, but just FYI if you are tired of lengthy solves (installing moose-dev using this instance of Conda will/is probably taking around 30 minutes), updating to a later release will greatly speed things up. To elaborate, your `conda info` should include this within the output:

```pre
         python version : 3.10.14.final.0
                 solver : libmamba (default)     <-------
       virtual packages : __archspec=1=m2
                          __conda=24.9.2=0
                          __osx=14.7.1=0
                          __unix=0=0
```


Anyways, the real issue here is that our packages are built using `conda-forge`'s compilers. So you'll need to add that channel into the mix in order to install anything from our `public` channel:

```bash
conda config --add channels conda-forge
conda config --add channels https://conda.software.inl.gov/public
```
_You'll get a warning about our channel already added but moving to the top... this is good_

Once that is complete, try your conda create command again!

**Reply by milljm** [reply]
I just read that this may be an instance of Conda you don't have control over (HPC). If that is true, I think in that case the following should work:

```bash
conda create -n testing moose-pyhit=2022 \
 --channel https://conda.software.inl.gov/public \
 --channel conda-forge
```

That way, you're not modifying the channels in place already.

**Reply by jvwilliams23** [reply]
Thanks @milljm, I managed to download `moose-pyhit`! Do you know how to get `moosetree`? My code relies on `moosetree.find`

**Reply by milljm** [reply, marked as answer]
Indeed, comes with :)
```python
>>> from pyhit import moosetree
>>> moosetree.__file__
'/Users/milljm/miniforge/envs/testing/lib/python3.10/site-packages/pyhit/moosetree/__init__.py'
```

**Reply by jvwilliams23** [reply]
Perfect, thank you!

**Reply by milljm** [reply]
Oh, sorry. I should have demonstrated `moosetree.find('node')`. But I was more interested in demonstrating access to moosetree, without needing a clone of the moose repository.

Cheers!
